Add Show Objective-C Selector code action

iMostfa opened the pull request:
Resolves https://github.com/swiftlang/sourcekit-lsp/issues/2145

Adds a *Show Objective-C Selector* code action for declarations exposed to Objective-C.

The action is only offered when the declaration under the cursor actually has a selector, and the selector is fetched while computing the code action, so invoking it needs no second request.

Requires https://github.com/swiftlang/swift/pull/86173.

https://github.com/user-attachments/assets/467cec3f-d0cb-4e64-91f1-adfc7879d33f

iMostfa committed:
Add Show Objective-C Selector code action

Adds a code action that reveals the Objective-C selector of a method that is
exposed to Objective-C, using the `source.request.objc.selector` sourcekitd
request added by https://github.com/swiftlang/swift/pull/86173.

The action is only offered when the declaration under the cursor actually has
a selector. Cursor info already reports the Objective-C USR (`c:`) for
declarations that are exposed to Objective-C -- via `@objc`, `@objcMembers` or
by overriding an Objective-C method -- and the Swift USR (`s:`) for everything
else, so that check costs nothing. The selector is fetched while computing the
code action and carried in the command, so invoking it needs no second request.

ahoppen reviewed (commented):
Thanks for opening the PR. One thing I’m not entirely convinced about is whether this should be a refactoring actions that modifies the current source file. I would probably have implemented this as a code action that shows a message with the Objective-C selector name to the user via the `window/showMessageRequest` request if the client supports that request. What do you think about that approach?

iMostfa commented:
https://github.com/user-attachments/assets/652f2408-ada2-4b61-9ae9-ed7ee8967a18

@ahoppen I tried to use `window/showMessageRequest` but i found that we won't provide any actions, So instead I used `window/showMessage` which works by just showing a message

 [e37ec29](https://github.com/swiftlang/sourcekit-lsp/pull/2399/commits/e37ec29e3086035ff9f477b2443bb2e49e9db6da)

After giving it more thought, I think it’s more useful to add the Objective-C selector as a comment rather than only showing it in a dialog.

The most common use case for this action is likely copying the selector to use it in an Objective-C file. While it’s possible to move the mouse and copy it from the dialog, it’s more convenient to stay on the keyboard and simply navigate to and select the comment.

Do you think it makes sense to prefer adding a comment in this case?

iMostfa commented:
fair points, so, if we decide to continue with just showMessage request, maybe we can rename the code action to be "Show Objective-c selector" instead of "Copy Objective-c selector" ?

iMostfa commented on Sources/SwiftLanguageService/SwiftLanguageService.swift:
```diff
@@ -962,13 +965,23 @@ extension SwiftLanguageService {
-    var refactorActions = cursorInfoResponse.refactorActions.compactMap {
-      let lspCommand = $0.asCommand()
+    var refactorActions: [CodeAction] = cursorInfoResponse.refactorActions.compactMap { action in
+      if action.actionString == "source.refactoring.kind.copy.objc.selector" {
```
Sorry for taking some time to check it, i checked both PRs, and currently this the content of each PR:

  #### Changes in swiftlang/swift#86173:
  - Fixed a bug in the request handler (inverted condition check)
  - Added `objc-selector` support to `sourcekitd-test` for testing
  - Moved test from `test/refactoring/` to `test/SourceKit/ObjCSelector/` since it's not a refactoring

  #### Changes here:
  - Removed dependency on the non-existent `source.refactoring.kind.copy.objc.selector` action
  - calling `source.request.objc.selector` directly and reads `key.text` from response

ahoppen commented on Sources/SwiftLanguageService/SwiftLanguageService.swift:
```diff
@@ -978,6 +979,22 @@ extension SwiftLanguageService {
+        return methodKinds.contains(kind)
+      }
+      return false
+    }
```
Does this mean that we show the refactoring action on every Swift method, even if it isn’t backed by an Objective-C method and thus the *Show Objective-C Selector* would just fail? Do we have some indication in the cursor info response already to see whether this method is defined in Objective-C / that we have an Objective-C selector? One ugly idea I have is checking whether the USR starts with `s:` but there might be nicer implementations as well.

iMostfa replied:
Good catch, yes it was offered on every function.

Your USR idea was the only option i found as well.

and is what the code does now: cursor info reports the Objective-C USR (`c:`) for declarations exposed to Objective-C and the Swift USR (`s:`) otherwise, so the check is free. We only ask `sourcekitd` for a selector when the cursor is on a function with a `c:` USR, and since that gives us the selector it's carried in the command, so invoking the action won't need a second request

wdyt ?

ahoppen reviewed (commented):
Nice, this looks good. Could you also add a test case for the new refactoring action? Both for the case where it shows and also a case checking that we don’t show it if the function doesn’t have an Objective-C selector.

iMostfa commented on Sources/SwiftLanguageService/ShowObjCSelectorCommand.swift:
```diff
@@ -0,0 +1,67 @@
+      let textDocument = TextDocumentIdentifier(fromLSPDictionary: documentDict)
+    else {
+      return nil
+    }
```
done

iMostfa commented on Sources/SKTestSupport/SkipUnless.swift:
```diff
@@ -459,6 +459,47 @@ package actor SkipUnless {
+    file: StaticString = #filePath,
+    line: UInt = #line
+  ) async throws {
+    return try await shared.skipUnlessSupportedByToolchain(swiftVersion: SwiftVersion(6, 5), file: file, line: line) {
```
i'm not sure if there's a better way to do this, i'm assuming that swift's PR will be in Swift 6.5, but i could be wrong here.

ahoppen commented on Sources/SKTestSupport/SkipUnless.swift:
```diff
@@ -459,6 +459,47 @@ package actor SkipUnless {
+        )
+
+        let testURL = scratchDirectory.appending(component: "test.swift")
+        try source.write(to: testURL, atomically: false, encoding: .utf8)
```
Could you use `writeWithRetry` here?

iMostfa replied:
Done in 1197fb47.

ahoppen commented on Sources/SwiftLanguageService/ShowObjCSelector.swift:
```diff
@@ -0,0 +1,40 @@
+  /// The Objective-C selector of the declaration at `position`, or `nil` if it isn't exposed to Objective-C or the
+  /// toolchain doesn't know the request.
+  func objcSelector(_ uri: DocumentURI, at position: Position) async -> String? {
+    return await orLog("Getting Objective-C selector", level: .debug) {
```
Why are you logging at the debug level here?

iMostfa replied:
It made sense when the request was sent for every
function and was expected to fail on most of them. Now that it only runs for
declarations cursor info already reports as exposed to Objective-C, a failure
is unexpected, so it uses the default level in 5da7c73c.

ahoppen commented on Sources/SwiftLanguageService/SwiftLanguageService.swift:
```diff
@@ -1084,6 +1084,23 @@ extension SwiftLanguageService {
       refactorActions.append(CodeAction(title: expandMacroCommand.title, kind: .refactor, command: expandMacroCommand))
     }
 
+    // Declarations exposed to Objective-C have an Objective-C USR (`c:`), everything else has a Swift USR (`s:`).
+    // Checking it avoids a semantic request per code action on Swift-only declarations.
+    let isOnObjCFunction = cursorInfoResponse.cursorInfo.contains { cursorInfo in
+      switch cursorInfo.symbolInfo.kind {
+      case .method, .function, .constructor:
+        return cursorInfo.symbolInfo.usr?.hasPrefix("c:") ?? false
+      default:
+        return false
+      }
+    }
+    if isOnObjCFunction, let selector = await objcSelector(params.textDocument.uri, at: params.range.lowerBound) {
+      let showObjCSelectorCommand = ShowObjCSelectorCommand(selector: selector).asCommand()
+      refactorActions.append(
+        CodeAction(title: showObjCSelectorCommand.title, kind: .refactor, command: showObjCSelectorCommand)
+      )
+    }
+
```
This isn’t really a refactoring action so the code action kind probably shouldn’t be `.refactor`. Could you check if we could either:
1. Define our own code action kind and announce it in the initialization options. This should be supported according to the LSP spec. Check that VS Code displays the code action.
> The set of kinds is open and client needs to announce the kinds it supports
> to the server during initialization.
2. If that doesn’t work, could we use `.empty` instead?

Also, the assumption that `retrieveRefactorCodeActions` only returns `.refactor` actions no longer holds, so the filter after `providersAndKinds` no longer works. I would suggest renaming this method from `retrieveRefactorCodeActions` to `retrieveCursorInfoBasedCodeActions` and make it take a list of requested code action kinds to then filter by them.

iMostfa replied:
thanks, i will need to dip up more to see which option is better

ahoppen commented on Sources/SwiftLanguageService/SwiftLanguageService.swift:
```diff
@@ -1084,6 +1084,23 @@ extension SwiftLanguageService {
+    let isOnObjCFunction = cursorInfoResponse.cursorInfo.contains { cursorInfo in
+      switch cursorInfo.symbolInfo.kind {
+      case .method, .function, .constructor:
+        return cursorInfo.symbolInfo.usr?.hasPrefix("c:") ?? false
```
Instead of interpreting the mangling prefix, could you inspect `key.decl_lang` from the cursor info response instead?

iMostfa replied:
as far as i can see , but it returns from [`getUIDForDeclLanguage`](https://github.com/swiftlang/swift/blob/d37faf0b09c2b819926f3a72549cf7bd5cf215a9/tools/SourceKit/lib/SwiftLang/SwiftLangSupport.cpp#L326-L330), which returns `source.lang.objc` only when the declaration has a Clang node:

```cpp
UIdentSwiftLangSupport::getUIDForDeclLanguage(const swift::Decl *D) {
  if (D->hasClangNode())
    return KindObjC;
  return KindSwift;
}
```

for `@objc func annotated()` i got `source.lang.swift`, only for `NSString.substring(from:)` it will work.

one thing we can do is try to check it first, otherwise we fallback for usr? but i'm not sure if makes any different. what do you think ?

ahoppen replied:
@rintaro @hamishknight Do you have a good suggestion of how to decide whether we should offer a code action to copy a Swift function’s Objective-C selector? Can you think of some property of an `AbstractFunctionDecl` that we could add to a cursor info response to indicate whether it’s callable from Objective-C?

ahoppen reviewed (commented):
Thanks for picking this back up.

iMostfa committed:
Use writeWithRetry in the Objective-C selector feature check

iMostfa committed:
Stop modelling Show Objective-C Selector as a refactoring

Showing a selector doesn't modify the source, so the action now uses the empty
code action kind instead of .refactor.

Since the provider no longer returns only refactorings, it can't be filtered by
a single kind in codeAction(_:). It is renamed to retrieveCursorInfoBasedCodeActions
and filters by the requested kinds itself. Matching is prefix aware so that a
refactor.extract action still satisfies a request for refactor, and the provider
returns early when the client asked for kinds it cannot produce, so that eg.
organize imports on save does not pay for a cursor info request.

iMostfa committed:
Log Objective-C selector request failures at the default level

The request is only sent for declarations that cursor info already reports as
exposed to Objective-C, so a failure is unexpected rather than routine.
